azure-iot-device/azure/iot/device/common/pipeline/pipeline_stages_mqtt.py

```diff
@@ -176,3 +210,32 @@ def _on_message_received(self, topic, payload):
             stage=self,
             event=pipeline_events_mqtt.IncomingMQTTMessageEvent(topic=topic, payload=payload),
         )
+
+    def _on_unexpected_connection(self):
```

Comment:
>_on_unexpected_connection [](start = 8, length = 25)

this should only get called when we get a successful CONACK but we weren't even trying to connect.  In this case, I think it makes sense to accept it as-is and call the on_connected handler,  it _shouldn't_ get called, but these 3 functions are all trying to make the best of an awkward situation.

Reply:
Are externally driven connects possible? Reconnects? If so, wouldn't this spuriously trigger? (my understanding is that this is intended to log unexpected connections as a result of threading issues in the pipeline)

---
In reply to: [302270895](https://github.com/Azure/azure-iot-sdk-python-preview/pull/136#discussion_r302270895) [](ancestors = 302270895)

Reply:
This should not be possible, but I believe strongly in handling unexpected conditions as gracefully as possible.  Regardless, this code has been refactored.

---
In reply to: [303137777](https://github.com/Azure/azure-iot-sdk-python-preview/pull/136#discussion_r303137777) [](ancestors = 303137777,302270895)